Source/Details/NSArray+Diffing.mm

```diff
@@ -0,0 +1,186 @@
+//
+//  NSArray+Diffing.m
+//  Texture
+//
+//  Copyright (c) 2014-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under the BSD-style license found in the
+//  LICENSE file in the /ASDK-Licenses directory of this source tree. An additional
+//  grant of patent rights can be found in the PATENTS file in the same directory.
+//
+//  Modifications to this file made after 4/13/2017 are: Copyright (c) 2017-present,
+//  Pinterest, Inc.  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+
+#import <AsyncDisplayKit/NSArray+Diffing.h>
+#import <unordered_map>
+#import <queue>
+#import <AsyncDisplayKit/ASAssert.h>
+
+@implementation NSArray (Diffing)
+
+typedef BOOL (^compareBlock)(id _Nonnull lhs, id _Nonnull rhs);
+
+- (void)asdk_diffWithArray:(NSArray *)array insertions:(NSIndexSet **)insertions deletions:(NSIndexSet **)deletions
+{
+  [self asdk_diffWithArray:array insertions:insertions deletions:deletions moves:nil compareBlock:[NSArray defaultCompareBlock]];
+}
+
+- (void)asdk_diffWithArray:(NSArray *)array insertions:(NSIndexSet **)insertions deletions:(NSIndexSet **)deletions
+              compareBlock:(compareBlock)comparison
+{
+  [self asdk_diffWithArray:array insertions:insertions deletions:deletions moves:nil compareBlock:comparison];
+}
+
+- (void)asdk_diffWithArray:(NSArray *)array insertions:(NSIndexSet **)insertions deletions:(NSIndexSet **)deletions
+                     moves:(NSArray<NSIndexPath *> **)moves
+{
+  [self asdk_diffWithArray:array insertions:insertions deletions:deletions moves:moves
+              compareBlock:[NSArray defaultCompareBlock]];
+}
+
+- (void)asdk_diffWithArray:(NSArray *)array insertions:(NSIndexSet **)insertions deletions:(NSIndexSet **)deletions
+                     moves:(NSArray<NSIndexPath *> **)moves compareBlock:(compareBlock)comparison
+{
+  typedef std::unordered_map<NSUInteger, std::queue<NSUInteger>> move_map;
+  NSAssert(comparison != nil, @"Comparison block is required");
+    NSAssert(moves == nil || comparison == [NSArray defaultCompareBlock], @"move detection requires isEqual: and hash (no custom compare)");
+  std::unique_ptr<move_map> potentialMoves(nullptr);
+  NSMutableArray<NSIndexPath *> *moveIndexPaths = nil;
+  NSMutableIndexSet *insertionIndexes = nil, *deletionIndexes = nil;
+  if (moves) {
+    potentialMoves = std::unique_ptr<move_map>(new move_map());
```

Comment:
Can you change this to stack allocation, matching the style of other unordered_map's in Texture? I think we can skip the typedef in that case too, as it is fairly clear in that style.

This will allocate it in some cases where it is not needed, but as a stack allocation, it is quite efficient.